pkg/download/downloader_test.go

```diff
@@ -37,24 +37,91 @@ func testdataPath() string {
 	return filepath.Join(pwd, "testdata")
 }
 
+type ArchiveTest struct {
+	in      string
+	files   []string
+	wantErr bool
+}
+
+func SymlinkTest1(ext string) ArchiveTest {
+	return ArchiveTest{
+		in: "test-with-symlinks" + ext,
+		files: []string{
+			"/symlinks/",
+			"/symlinks/message",
+			"/symlinks/msg",
+		},
+		wantErr: false,
+	}
+}
+
+func SymlinkTest2(ext string) ArchiveTest {
+	return ArchiveTest{
+		in: "test-with-symlinks-escaping-parent" + ext,
+		files: []string{
+			"/escaping-link-test2/",
+			"/escaping-link-test2/yo/",
+			"/escaping-link-test2/yo/mo/",
+			"/escaping-link-test2/yo/mo/zoo",
+			"/escaping-link-test2/z",
+			"/escaping-link-test2/zz",
+			"/escaping-link-test2/zzz", // this escapes only to the staging area, so wantErr: false
+		},
+		wantErr: false,
+	}
+}
+
+func SymlinkTest3(ext string) ArchiveTest {
+	return ArchiveTest{
+		in: "test-with-symlinks-escaping-parent2" + ext,
+		files: []string{
+			"/escaping-link-test3/",
+			"/escaping-link-test3/baz", // this escapes only to the staging area, so wantErr: false
+			"/escaping-link-test3/foo/",
+			"/escaping-link-test3/foo/bar/",
+		},
+		wantErr: false,
+	}
+}
+
+func SymlinkTest4(ext string) ArchiveTest {
+	return ArchiveTest{
+		in:      "test-with-symlinks-escaping-parent3" + ext,
+		files:   nil,
+		wantErr: true,
+	}
+}
+
+func SymlinkTest5(ext string) ArchiveTest {
+	return ArchiveTest{
+		in:      "test-with-symlinks-escaping-absolute" + ext,
+		files:   nil,
+		wantErr: true,
+	}
+}
+
 func Test_extractZIP(t *testing.T) {
-	tests := []struct {
-		in    string
-		files []string
-	}{
+	tests := []ArchiveTest{
 		{
 			in: "test-with-directory.zip",
 			files: []string{
 				"/test/",
 				"/test/foo",
 			},
+			wantErr: false,
```

Comment:
as you can tell, these tests were designed to succeed. for "error" cases, you can test them in unit-tests.